TESTS/mbedmicro-rtos-mbed/rtostimer/main.cpp

```diff
@@ -0,0 +1,355 @@
+#error invalid TEST_RESTART_DELAY_MS value
+#endif
+
+void timer_callback(void const *arg)
+{
+    Semaphore *sem = (Semaphore *) arg;
+    sem->release();
+}
+
+/** Test one-shot not restarted when elapsed
+ *
+ * Given a one-shot timer
+ * When the timer is started
+ *     and given time elapses
+ * Then timer stops
+ *     and elapsed time matches given delay
+ *     and timer stays stopped
+ */
+void test_oneshot_not_restarted()
+{
+    Semaphore sem(1);
+    RtosTimer timer(mbed::callback(timer_callback, (void const *) &sem), osTimerOnce);
+    osStatus stat = timer.stop();
+    TEST_ASSERT_EQUAL(osErrorResource, stat);
+
+    sem.wait(0);
+    stat = timer.start(TEST_DELAY_MS);
+    uint32_t start = us_ticker_read();
+    TEST_ASSERT_EQUAL(osOK, stat);
+
+    int32_t slots = sem.wait(TEST_DELAY_MS + 1);
+    uint32_t end = us_ticker_read();
+    TEST_ASSERT_EQUAL(1, slots);
+    TEST_ASSERT_UINT32_WITHIN(TEST_DELAY_MS_DELTA * 1000, TEST_DELAY_MS * 1000, end - start);
+
+    slots = sem.wait(TEST_DELAY_MS + 1);
+    TEST_ASSERT_EQUAL(0, slots);
+
+    stat = timer.stop();
+    TEST_ASSERT_EQUAL(osErrorResource, stat);
+}
+
+/** Test periodic repeats continuously
+ *
+ * Given a periodic timer
+ * When timer is started
+ *     and given time elapses
+ * Then timer repeats its operation
+ * When timer is stopped
+ * Then timer stops operation
+ */
+void test_periodic_repeats()
+{
+    Semaphore sem(1);
+    RtosTimer timer(mbed::callback(timer_callback, (void const *) &sem), osTimerPeriodic);
+    osStatus stat = timer.stop();
+    TEST_ASSERT_EQUAL(osErrorResource, stat);
+
+    sem.wait(0);
+    stat = timer.start(TEST_DELAY_MS);
+    uint32_t t1 = us_ticker_read();
+    TEST_ASSERT_EQUAL(osOK, stat);
+
+    int32_t slots = sem.wait(TEST_DELAY_MS + 1);
+    uint32_t t2 = us_ticker_read();
+    TEST_ASSERT_EQUAL(1, slots);
+    TEST_ASSERT_UINT32_WITHIN(TEST_DELAY_MS_DELTA * 1000, TEST_DELAY_MS * 1000, t2 - t1);
+
+    slots = sem.wait(TEST_DELAY_MS + 1);
+    uint32_t t3 = us_ticker_read();
+    TEST_ASSERT_EQUAL(1, slots);
+    TEST_ASSERT_UINT32_WITHIN(TEST_DELAY_MS_DELTA * 1000, TEST_DELAY_MS * 1000, t3 - t2);
+
+    stat = timer.stop();
+    TEST_ASSERT_EQUAL(osOK, stat);
+
+    slots = sem.wait(TEST_DELAY_MS + 1);
+    TEST_ASSERT_EQUAL(0, slots);
+
+    stat = timer.stop();
+    TEST_ASSERT_EQUAL(osErrorResource, stat);
+}
+
+/** Test timer can be restarted
+ *
+ * Given a one-shot timer
+ * When the timer is started
+ *     and @a start is called again before given time elapses
+ *     and given time elapses
+ * Then timer stops
+ *     and elapsed time is greater than original delay
+ */
+void test_restart()
+{
+    Semaphore sem(1);
+    RtosTimer timer(mbed::callback(timer_callback, (void const *) &sem), osTimerOnce);
+    osStatus stat = timer.stop();
+    TEST_ASSERT_EQUAL(osErrorResource, stat);
+
+    sem.wait(0);
+    stat = timer.start(TEST_DELAY_MS);
+    uint32_t t1 = us_ticker_read();
```

Comment:
I would recomend using Timer class instead of us ticker read to capture the duration (C++ classes should provide all its needed)

Reply:
Sure, I'll update time measurement with Timers.